edgedb-client/src/client.rs

```diff
@@ -142,15 +149,62 @@ impl Connection {
         return &self.version
     }
     pub async fn passive_wait<T>(&mut self) -> T {
-        let mut buf = [0u8; 1];
-        self.input.read(&mut buf[..]).await.ok();
+        let (_, mut reader, _) = self.split();
+        reader.passive_wait().await.ok();
         // any erroneous or successful read (even 0) means need reconnect
-        self.dirty = true;
+        self.state = State::Dirty;
         pending::<()>().await;
         unreachable!();
     }
+    async fn do_pings(&mut self) -> Result<(), Error> {
+        use async_std::io;
+
+        let ivl = self.ping_interval;
+        let (mut writer, mut reader, state) = self.split();
+
+        while *state == State::Normal {
+
+            *state = State::Dirty;
+            writer.send_messages(&[ClientMessage::Sync]).await?;
+            *state = State::AwaitingPing;
+            reader.wait_ready().await?;
+            *state = State::Normal;
```

Comment:
If `wait_ready()` fails, should we set the state to dirty?

Reply:
Yes. Good catch!